Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -0,0 +1,841 @@
+            }
+
+            var reanalyzeEntries = aliases.SelectMany(a => _analyzer.GetEntriesThatImportModule(a, true)).ToArray();
+            var first = aliases.FirstOrDefault();
+
+            var pyItem = _analyzer.AddModule(first, path, documentUri, cookie);
+            item = pyItem;
+            foreach (var a in aliases.Skip(1)) {
+                _analyzer.AddModuleAlias(first, a);
+            }
+
+            var actualItem = GetOrAddEntry(documentUri, item);
+            if (actualItem != item) {
+                return Task.FromResult(actualItem);
+            }
+
+            if (_clientCaps?.python?.analysisUpdates ?? false) {
+                pyItem.OnNewAnalysis += ProjectEntry_OnNewAnalysis;
+            }
+
+            if (item is IDocument doc) {
+                EnqueueItem(doc);
+            }
+
+            if (reanalyzeEntries != null) {
+                foreach (var entryRef in reanalyzeEntries) {
+                    _queue.Enqueue(entryRef, AnalysisPriority.Low);
+                }
+            }
+
+            return Task.FromResult(item);
+        }
+
+        private static bool IsDocumentChanged(IDocument doc, VersionCookie previousResult) {
+            if (previousResult == null) {
+                return true;
+            }
+
+            int seen = 0;
+            foreach (var part in doc.DocumentParts) {
+                if (!previousResult.Versions.TryGetValue(part, out var bv)) {
+                    return true;
+                }
+                if (doc.GetDocumentVersion(part) > bv.Version) {
+                    return true;
+                }
+                seen += 1;
+            }
+            if (seen != doc.DocumentParts.Count()) {
+                return true;
+            }
+            return false;
+        }
+
+        private IDisposable GetDocumentParseCounter(IDocument doc, out int count) {
+            VolatileCounter counter;
+            lock (_pendingParse) {
+                if (!_pendingParse.TryGetValue(doc, out counter)) {
+                    _pendingParse[doc] = counter = new VolatileCounter();
+                    // Automatically remove counter from the dictionary when it
+                    // reaches zero.
+                    counter.WaitForZeroAsync().ContinueWith(t => {
+                        if (t.IsCompleted) {
+                            lock (_pendingParse) {
+                                if (counter.IsZero && _pendingParse.TryGetValue(doc, out var existing) && existing == counter) {
+                                    _pendingParse.Remove(doc);
+                                }
+                            }
+                        }
+                    });
+                }
+                var res = counter.Incremented();
+                count = counter.Count;
+                return res;
+            }
+        }
+
+        private async void EnqueueItem(IDocument doc, AnalysisPriority priority = AnalysisPriority.Normal) {
```

Comment:
`async void`

Reply:
Correct. This is fire-and-forget by design.